File: ForgeHub/server/Api/AdminApi.cs
using ForgeHub.Auth;
using ForgeHub.Data;

namespace ForgeHub.Api;

public static class AdminApi
{
    public static void MapAdmin(this WebApplication app)
    {
        // --------------------------------------------------------------
        app.MapGet("/api/tenants", (HttpContext context) =>
        {
            using var connection = context.Db(out var tenantId);

            var rows = Sql.Rows(connection,
                @"SELECT t.id, t.name, t.display_name, t.created_at,
                         (SELECT COUNT(*) FROM users u WHERE u.tenant_id = t.id) AS user_count,
                         (SELECT COUNT(*) FROM robots r WHERE r.tenant_id = t.id) AS robot_count
                    FROM tenants t
                   ORDER BY t.name");

            return Results.Ok(rows);
        });

        // --------------------------------------------------------------
        // Ringkasan pengguna. Ringkasan kata sandi TIDAK pernah ikut, bahkan
        // dalam bentuk teracaknya: ringkasan yang bocor masih bisa ditebak
        // di luar sini tanpa batas percobaan.
        app.MapGet("/api/users", (HttpContext context) =>
        {
            using var connection = context.Db(out var tenantId);

            var rows = Sql.Rows(connection,
                @"SELECT id, username, display_name, email, role, is_active, created_at, last_login_at
                    FROM users
                   WHERE tenant_id = @p0
                   ORDER BY username",
                tenantId);

            return Results.Ok(rows);
        });

        app.MapPost("/api/users", async (HttpContext context) =>
        {
            if (!IsAdministrator(context))
                return Forbidden("Hanya Administrator yang boleh mengubah pengguna.");

            var body = await context.Body();

            var username = body.Str("username");
            var password = body.Str("password");

            if (string.IsNullOrWhiteSpace(username))
                return Results.BadRequest(new { error = "Nama pengguna wajib diisi." });

            using var connection = context.Db(out var tenantId);

            var existing = (string)Sql.Scalar(connection,
                "SELECT id FROM users WHERE tenant_id = @p0 AND username = @p1", tenantId, username);

            if (existing != null)
            {
                Sql.Exec(connection,
                    @"UPDATE users
                         SET display_name = COALESCE(@p0, display_name),
                             email = COALESCE(@p1, email),
                             role = COALESCE(@p2, role),
                             is_active = @p3
                       WHERE id = @p4",
                    body.Str("displayName"), body.Str("email"), body.Str("role"),
                    body.Flag("isActive", true) ? 1 : 0, existing);

                // Kata sandi hanya diganti kalau memang dikirim. Tanpa syarat ini,
                // menyunting alamat surel akan diam-diam mengosongkan sandinya.
                if (!string.IsNullOrEmpty(password))
                {
                    Sql.Exec(connection, "UPDATE users SET password_hash = @p0 WHERE id = @p1",
                        Passwords.Hash(password), existing);
                }

                return Results.Ok(new { ok = true, created = false });
            }

            if (string.IsNullOrEmpty(password) || password.Length < 6)
                return Results.BadRequest(new { error = "Pengguna baru butuh kata sandi minimal 6 karakter." });

            Sql.Exec(connection,
                @"INSERT INTO users (id, tenant_id, username, password_hash, display_name, email, role, is_active, created_at)
                  VALUES (@p0, @p1, @p2, @p3, @p4, @p5, @p6, @p7, @p8)",
                Sql.NewId(), tenantId, username, Passwords.Hash(password),
                body.Str("displayName", username), body.Str("email"),
                body.Str("role", "Automation User"), body.Flag("isActive", true) ? 1 : 0, Sql.Now());

            return Results.Ok(new { ok = true, created = true });
        });

        app.MapDelete("/api/users/{username}", (HttpContext context, string username) =>
        {
            if (!IsAdministrator(context))
                return Forbidden("Hanya Administrator yang boleh menghapus pengguna.");

            // Menghapus diri sendiri akan mengunci orangnya keluar dari ForgeHub
            // miliknya sendiri, dan tidak ada jalan masuk lain untuk membatalkannya.
            if (string.Equals(username, context.User()?.Username, StringComparison.OrdinalIgnoreCase))
                return Results.BadRequest(new { error = "Tidak bisa menghapus akun yang sedang dipakai." });

            using var connection = context.Db(out var tenantId);

            // Penyewa tanpa satu pun administrator aktif tidak bisa diurus lagi.
            var admins = Sql.Count(connection,
                "SELECT COUNT(*) FROM users WHERE tenant_id = @p0 AND role = 'Administrator' AND is_active = 1",
                tenantId);

            var isAdmin = Sql.Count(connection,
                "SELECT COUNT(*) FROM users WHERE tenant_id = @p0 AND username = @p1 AND role = 'Administrator'",
                tenantId, username) > 0;

            if (isAdmin && admins <= 1)
                return Results.BadRequest(new { error = "Ini satu-satunya Administrator yang tersisa." });

            var removed = Sql.Exec(connection,
                "DELETE FROM users WHERE tenant_id = @p0 AND username = @p1", tenantId, username);

            return removed == 0 ? Results.NotFound(new { error = "Pengguna tidak ada." }) : Results.Ok(new { ok = true });
        });

        // --------------------------------------------------------------
        app.MapGet("/api/roles", (HttpContext context) =>
        {
            using var connection = context.Db(out var tenantId);

            var rows = Sql.Rows(connection,
                @"SELECT r.id, r.name, r.description, r.permissions, r.created_at,
                         (SELECT COUNT(*) FROM users u
                           WHERE u.tenant_id = r.tenant_id AND u.role = r.name) AS user_count
                    FROM roles r
                   WHERE r.tenant_id = @p0
                   ORDER BY r.name",
                tenantId);

            return Results.Ok(rows);
        });

        // --------------------------------------------------------------
        app.MapGet("/api/licensing", (HttpContext context) =>
        {
            using var connection = context.Db(out var tenantId);

            var rows = Sql.Rows(connection,
                "SELECT id, product, total, used, expires_at FROM licenses WHERE tenant_id = @p0 ORDER BY product",
                tenantId);

            // "Terpakai" dihitung dari robot yang benar-benar ada, bukan dari
            // angka yang pernah dituliskan seseorang ke kolom used.
            var attended = Sql.Count(connection,
                "SELECT COUNT(*) FROM robots WHERE tenant_id = @p0 AND type = 'Attended'", tenantId);

            var unattended = Sql.Count(connection,
                "SELECT COUNT(*) FROM robots WHERE tenant_id = @p0 AND type <> 'Attended'", tenantId);

            foreach (var row in rows)
            {
                var product = (string)row["product"] ?? "";
                row["used"] = product.Contains("Attended") && !product.Contains("Unattended")
                    ? attended : unattended;
            }

            return Results.Ok(rows);
        });

        // --------------------------------------------------------------
        app.MapGet("/api/settings", (HttpContext context) =>
        {
            using var connection = context.Db(out var tenantId);

            return Results.Ok(new
            {
                tenant = context.User()?.TenantName,
                serverTime = Sql.Now(),
                machineName = System.Environment.MachineName,
                robotOfflineAfterSeconds = RobotsApi.OfflineAfterSeconds,
                tokenLifetimeHours = (int)Tokens.Lifetime.TotalHours,
                dataDirectory = Program.DataDirectory,
                counts = new
                {
                    users = Sql.Count(connection, "SELECT COUNT(*) FROM users WHERE tenant_id = @p0", tenantId),
                    robots = Sql.Count(connection, "SELECT COUNT(*) FROM robots WHERE tenant_id = @p0", tenantId),
                    processes = Sql.Count(connection, "SELECT COUNT(*) FROM processes WHERE tenant_id = @p0", tenantId),
                    jobs = Sql.Count(connection, "SELECT COUNT(*) FROM jobs WHERE tenant_id = @p0", tenantId),
                    logs = Sql.Count(connection, "SELECT COUNT(*) FROM logs WHERE tenant_id = @p0", tenantId),
                },
            });
        });
    }

    // ------------------------------------------------------------------

    private static bool IsAdministrator(HttpContext context) =>
        context.User()?.Role == "Administrator";

    private static IResult Forbidden(string message) =>
        Results.Json(new { error = message }, statusCode: StatusCodes.Status403Forbidden);
}

File: ForgeHub/backend/src/main/java/id/jakforge/forgehub/api/AdminApi.java
package id.jakforge.forgehub.api;

import id.jakforge.forgehub.data.Db;
import id.jakforge.forgehub.security.CurrentUser;
import id.jakforge.forgehub.security.ForgeHubPrincipal;
import id.jakforge.forgehub.security.JwtService;
import id.jakforge.forgehub.security.Passwords;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Pengguna, peran, penyewa, lisensi, dan setelan.
 *
 * <p>Semuanya hanya-baca kecuali pengguna, dan perubahan pengguna dibatasi ke
 * Administrator. Pembatasannya diperiksa di sini, bukan diserahkan ke
 * antarmuka: tombol yang disembunyikan tetap bisa dilewati oleh siapa pun yang
 * memanggil API-nya langsung.
 */
@RestController
@RequestMapping("/api")
public class AdminApi {

    private static final int PUTUS_SETELAH_DETIK = 45;

    private final Db db;
    private final JwtService jwt;
    private final String zonaTampilan;

    public AdminApi(Db db, JwtService jwt,
                    @Value("${forgehub.display-timezone:UTC}") String zonaTampilan) {
        this.db = db;
        this.jwt = jwt;
        this.zonaTampilan = zonaTampilan;
    }

    // -----------------------------------------------------------------
    // Penyewa
    // -----------------------------------------------------------------

    @GetMapping("/tenants")
    public List<Map<String, Object>> penyewa() {
        return db.rows("""
                SELECT t.id, t.name, t.display_name, t.created_at,
                       (SELECT count(*) FROM users u WHERE u.tenant_id = t.id)  AS user_count,
                       (SELECT count(*) FROM robots r WHERE r.tenant_id = t.id) AS robot_count
                  FROM tenants t
                 ORDER BY t.name
                """);
    }

    // -----------------------------------------------------------------
    // Pengguna
    // -----------------------------------------------------------------

    /**
     * Ringkasan pengguna.
     *
     * <p>Kolom {@code password_hash} TIDAK ikut dipilih, bahkan dalam bentuk
     * teracaknya: ringkasan yang bocor masih bisa ditebak di luar sini tanpa
     * batas percobaan dan tanpa ada yang tahu.
     */
    @GetMapping("/users")
    public List<Map<String, Object>> pengguna() {
        return db.rows("""
                SELECT id, username, display_name, email, role, is_active, created_at, last_login_at
                  FROM users
                 WHERE tenant_id = ?
                 ORDER BY username
                """, CurrentUser.get().tenantId());
    }

    @PostMapping("/users")
    @Transactional
    public ResponseEntity<?> simpanPengguna(@RequestBody(required = false) Map<String, Object> body) {
        ForgeHubPrincipal p = CurrentUser.get();

        if (!administrator(p)) return terlarang("Hanya Administrator yang boleh mengubah pengguna.");

        String username = Badan.nama(body, "username");

        if (username == null) {
            return ResponseEntity.badRequest().body(Map.of("error", "Nama pengguna wajib diisi."));
        }

        String password = Badan.teks(body, "password");
        UUID tenantId = p.tenantId();

        Object adaId = db.scalar("SELECT id FROM users WHERE tenant_id = ? AND username = ?",
                tenantId, username);

        if (adaId != null) {
            db.exec("""
                    UPDATE users
                       SET display_name = COALESCE(?, display_name),
                           email = COALESCE(?, email),
                           role = COALESCE(?, role),
                           is_active = ?
                     WHERE tenant_id = ? AND username = ?
                    """, Badan.teks(body, "displayName"), Badan.teks(body, "email"),
                    Badan.teks(body, "role"), Badan.benar(body, "isActive", true),
                    tenantId, username);

            // Kata sandi hanya diganti kalau memang dikirim. Tanpa syarat ini,
            // menyunting alamat surel akan diam-diam mengosongkan sandinya —
            // dan yang bersangkutan baru tahu saat gagal masuk besok pagi.
            if (password != null && !password.isEmpty()) {
                if (password.length() < 6) {
                    return ResponseEntity.badRequest().body(Map.of(
                            "error", "Kata sandi minimal 6 karakter."));
                }

                db.exec("UPDATE users SET password_hash = ? WHERE tenant_id = ? AND username = ?",
                        Passwords.hash(password), tenantId, username);
            }

            Map<String, Object> hasil = new LinkedHashMap<>();
            hasil.put("ok", true);
            hasil.put("created", false);

            return ResponseEntity.ok(hasil);
        }

        if (password == null || password.length() < 6) {
            return ResponseEntity.badRequest().body(Map.of(
                    "error", "Pengguna baru butuh kata sandi minimal 6 karakter."));
        }

        db.exec("""
                INSERT INTO users (id, tenant_id, username, password_hash, display_name,
                                   email, role, is_active, created_at)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, now())
                """, Db.newId(), tenantId, username, Passwords.hash(password),
                Badan.teks(body, "displayName", username), Badan.teks(body, "email"),
                Badan.teks(body, "role", "Automation User"), Badan.benar(body, "isActive", true));

        Map<String, Object> hasil = new LinkedHashMap<>();
        hasil.put("ok", true);
        hasil.put("created", true);

        return ResponseEntity.ok(hasil);
    }

    @DeleteMapping("/users/{username}")
    @Transactional
    public ResponseEntity<?> hapusPengguna(@PathVariable String username) {
        ForgeHubPrincipal p = CurrentUser.get();

        if (!administrator(p)) return terlarang("Hanya Administrator yang boleh menghapus pengguna.");

        // Menghapus diri sendiri akan mengunci orangnya keluar dari ForgeHub
        // miliknya sendiri, dan tidak ada jalan masuk lain untuk membatalkannya.
        if (username.equalsIgnoreCase(p.username())) {
            return ResponseEntity.badRequest().body(Map.of(
                    "error", "Tidak bisa menghapus akun yang sedang dipakai."));
        }

        UUID tenantId = p.tenantId();

        // Penyewa tanpa satu pun administrator aktif tidak bisa diurus lagi:
        // tidak ada yang bisa membuat administrator baru, dan tidak ada pintu
        // belakang untuk memperbaikinya.
        boolean iniAdmin = db.exists("""
                SELECT count(*) FROM users
                 WHERE tenant_id = ? AND username = ? AND role = 'Administrator'
                """, tenantId, username);

        long jumlahAdmin = db.count("""
                SELECT count(*) FROM users
                 WHERE tenant_id = ? AND role = 'Administrator' AND is_active
                """, tenantId);

        if (iniAdmin && jumlahAdmin <= 1) {
            return ResponseEntity.badRequest().body(Map.of(
                    "error", "Ini satu-satunya Administrator yang tersisa."));
        }

        int terhapus = db.exec("DELETE FROM users WHERE tenant_id = ? AND username = ?",
                tenantId, username);

        return terhapus == 0
                ? ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("error", "Pengguna tidak ada."))
                : ResponseEntity.ok(Map.of("ok", true));
    }

    // -----------------------------------------------------------------
    // Peran dan lisensi
    // -----------------------------------------------------------------

    @GetMapping("/roles")
    public List<Map<String, Object>> peran() {
        return db.rows("""
                SELECT r.id, r.name, r.description, r.permissions, r.created_at,
                       (SELECT count(*) FROM users u
                         WHERE u.tenant_id = r.tenant_id AND u.role = r.name) AS user_count
                  FROM roles r
                 WHERE r.tenant_id = ?
                 ORDER BY r.name
                """, CurrentUser.get().tenantId());
    }

    /**
     * Lisensi.
     *
     * <p>"Terpakai" dihitung dari robot yang benar-benar ada, bukan dari angka
     * yang pernah dituliskan seseorang ke kolom {@code used}. Angka yang
     * disimpan akan menyimpang begitu satu robot dihapus tanpa lewat layar ini.
     */
    @GetMapping("/licensing")
    public List<Map<String, Object>> lisensi() {
        UUID tenantId = CurrentUser.get().tenantId();

        List<Map<String, Object>> baris = db.rows("""
                SELECT id, product, total, used, expires_at
                  FROM licenses WHERE tenant_id = ? ORDER BY product
                """, tenantId);

        long attended = db.count(
                "SELECT count(*) FROM robots WHERE tenant_id = ? AND type = 'Attended'", tenantId);

        long unattended = db.count(
                "SELECT count(*) FROM robots WHERE tenant_id = ? AND type <> 'Attended'", tenantId);

        for (Map<String, Object> b : baris) {
            String produk = String.valueOf(b.get("product"));

            b.put("used", produk.contains("Attended") && !produk.contains("Unattended")
                    ? attended : unattended);
        }

        return baris;
    }

    // -----------------------------------------------------------------
    // Setelan
    // -----------------------------------------------------------------

    @GetMapping("/settings")
    public Map<String, Object> setelan() {
        ForgeHubPrincipal p = CurrentUser.get();

        Map<String, Object> jumlah = db.row("""
                SELECT (SELECT count(*) FROM users     WHERE tenant_id = ?) AS users,
                       (SELECT count(*) FROM robots    WHERE tenant_id = ?) AS robots,
                       (SELECT count(*) FROM processes WHERE tenant_id = ?) AS processes,
                       (SELECT count(*) FROM jobs      WHERE tenant_id = ?) AS jobs,
                       (SELECT count(*) FROM logs      WHERE tenant_id = ?) AS logs
                """, p.tenantId(), p.tenantId(), p.tenantId(), p.tenantId(), p.tenantId());

        Object namaPenyewa = db.scalar("SELECT name FROM tenants WHERE id = ?", p.tenantId());

        Map<String, Object> hasil = new LinkedHashMap<>();
        hasil.put("tenant", namaPenyewa);
        hasil.put("serverTime", Db.nowText());
        hasil.put("displayTimezone", zonaTampilan);
        hasil.put("robotOfflineAfterSeconds", PUTUS_SETELAH_DETIK);
        hasil.put("tokenLifetimeHours", jwt.getExpirationMinutes() / 60);

        // "database", bukan "dataDirectory": di sini datanya ada di PostgreSQL,
        // bukan di sebuah folder. Menyisakan nama medan yang lama akan membuat
        // layar Settings menampilkan jalur folder yang tidak berarti apa-apa.
        hasil.put("database", "PostgreSQL");
        hasil.put("counts", jumlah);

        return hasil;
    }

    // -----------------------------------------------------------------

    private static boolean administrator(ForgeHubPrincipal p) {
        return "Administrator".equals(p.role());
    }

    private static ResponseEntity<?> terlarang(String pesan) {
        return ResponseEntity.status(HttpStatus.FORBIDDEN).body(Map.of("error", pesan));
    }
}
